### PR TITLE
[RFC][PL-468] Create an entrypoint for graylog

### PR DESCRIPTION
**Pull request by raffo0707**
https://uniplaces.atlassian.net/browse/PL-468

### CODE REVIEW

File: cmd/bulk.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/uniplaces/logfairy/dto/dashboard"
+	"github.com/uniplaces/logfairy/dto/stream"
+	"github.com/uniplaces/logfairy/infrastructure/api"
+)
+
+// bulkCmd represents the bulk command
+var bulkCmd = &cobra.Command{
+	Use:   "bulk",
+	Short: "bulk allows to create a set of streams, dashboards and widgets",
+	Long: `bulk will try to create all the definitions in the config file.
```
Comment (hfcorreia) [inline cmd/bulk.go]: export the long to a const

File: cmd/bulk.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      ]
+    }
+  ]
+}
+
+Please notice that when creating dashboards widgets are not allowed, 
+behind the scene the widget are extracted from the dashboard structure.
+Also, note that the value of the field "stream_id" in the widget object
+reffer to the stream position in the stream list.`,
+	Run: func(cmd *cobra.Command, args []string) {
```
Comment (hfcorreia) [inline cmd/bulk.go]: extract to: `func Run(cmd *cobra.Command, args []string) {`
then 
```
{
 ...
 Run: run,
}
```

File: cmd/bulk.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		definition, err := ioutil.ReadFile(definitions)
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
```
Comment (hfcorreia) [inline cmd/bulk.go]: Use `log.Fatalf()` instead of `fmt` and `os`

File: cmd/bulk.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+		}
+	},
+}
+
+func init() {
+	widgetClient = api.NewWidgetClient(graylog)
+	dashboardClient = api.NewDashboardClient(graylog)
+	streamClient = api.NewStreamClient(graylog)
+
+	bulkCmd.Flags().StringVarP(&definitions, "config", "c", "", "config file containing the definitions")
```
Comment (hfcorreia) [inline cmd/bulk.go]: change to this format:
```
bulkCmd.
  Flags().
  StringVarP(&definitions, "config", "c", "", "config file containing the definitions")
```

File: cmd/bulk.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            "trend": true,
+            "query": "foo bar"
+          }
+        }
+      ]
+    }
+  ]
+}
+
+Please notice that when creating dashboards widgets are not allowed,
```
Comment (jkrnak) [inline cmd/bulk.go]: Can you please rephrase this sentence? 
I'm not sure what is it trying to get my attention to.

File: cmd/bulk.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	},
+}
+
+func init() {
+	widgetClient = api.NewWidgetClient(graylog)
+	dashboardClient = api.NewDashboardClient(graylog)
+	streamClient = api.NewStreamClient(graylog)
+
+	bulkCmd.Flags().StringVarP(&definitions, "config", "c", "", "config file containing the definitions")
+	bulkCmd.MarkFlagRequired("config")
```
Comment (hfcorreia) [inline cmd/bulk.go]: this can be appended to the previous statement

File: cmd/bulk.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        }
+      ]
+    }
+  ]
+}
+
+Please notice that when creating dashboards widgets are not allowed, 
+behind the scene the widget are extracted from the dashboard structure.
+Also, note that the value of the field "stream_id" in the widget object
+reffer to the stream position in the stream list.`,
```
Comment (jkrnak) [inline cmd/bulk.go]: "referr"

File: cmd/bulk.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+func createWidget(widgetToCreate dashboard.Widget, dashboardID string, streams []stream.Stream) error {
+	streamPosition, err := strconv.Atoi(widgetToCreate.Config.StreamID)
+	if err != nil {
+		return err
+	}
+
+	widgetToCreate.Config.StreamID = *streams[streamPosition].ID
+	_, err = widgetClient.Create(widgetToCreate, dashboardID)
```
Comment (hfcorreia) [inline cmd/bulk.go]: get the return value, and print out as info

File: cmd/bulk.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	return err
+}
+
+func extractWidgets(dashboardToCreate *dashboard.Dashboard) []dashboard.Widget {
+	if dashboardToCreate.Widgets == nil {
+		return []dashboard.Widget{}
+	}
+
+	widgets := dashboardToCreate.Widgets
+	dashboardToCreate.Widgets = nil
```
Comment (hfcorreia) [inline cmd/bulk.go]: do you really need to remove this from the argument?

File: cmd/dashboard.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+		prettyPrint(dashboard)
+	},
+}
+
+// CreateDashboardCmd represents the create command
+var CreateDashboardCmd = &cobra.Command{
+	Use:   "create",
+	Short: "crate a dashboard",
+	Long: `dashboard create will try to create the dashboard defined in the config file.
```
Comment (hfcorreia) [inline cmd/dashboard.go]: extract the long

File: cmd/dashboard.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	
+The expected json is:
+{
+  "title": "bar",
+  "description": "description for bar dashboard"
+}`,
+	Run: func(cmd *cobra.Command, args []string) {
+		definition, err := ioutil.ReadFile(definitions)
+		if err != nil {
+			fmt.Println(err)
```
Comment (hfcorreia) [inline cmd/dashboard.go]: fatal

File: cmd/dashboard.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+	Use:   "create",
+	Short: "crate a dashboard",
+	Long: `dashboard create will try to create the dashboard defined in the config file.
+	
+The expected json is:
+{
+  "title": "bar",
+  "description": "description for bar dashboard"
+}`,
+	Run: func(cmd *cobra.Command, args []string) {
```
Comment (hfcorreia) [inline cmd/dashboard.go]: i would extract the func

File: cmd/dashboard.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	Short: "handle dashboards actions",
+	Long:  ``,
+}
+
+// ListDashboardsCmd represents the List command
+var ListDashboardsCmd = &cobra.Command{
+	Use:   "list",
+	Short: "list dashboards",
+	Long:  `dashboard list look up for all the dashboards it is allowed to reach.`,
+	Run: func(cmd *cobra.Command, args []string) {
```
Comment (hfcorreia) [inline cmd/dashboard.go]: i would extract the func

File: cmd/dashboard.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+		prettyPrint(dashboards)
+	},
+}
+
+// GetDashboardCmd represents the list single dashboard command
+var GetDashboardCmd = &cobra.Command{
+	Use:   "get",
+	Short: "list a single dashboard",
+	Long:  `dashboard get look up for a dashboard with the given dashboard id and print out the json.`,
+	Run: func(cmd *cobra.Command, args []string) {
```
Comment (hfcorreia) [inline cmd/dashboard.go]: i would extract the func

File: cmd/dashboard.go [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+		fmt.Printf("dashboard id: %s", dashboardID)
+	},
+}
+
+func init() {
+	initConfig()
+
+	dashboardClient = api.NewDashboardClient(graylog)
+
+	GetDashboardCmd.Flags().StringVarP(&dashboardID, "dashboard_id", "d", "", "id of dashboard to find")
```
Comment (hfcorreia) [inline cmd/dashboard.go]: change the format, check comment above

File: cmd/dashboard.go [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+	GetDashboardCmd.MarkFlagRequired("dashboard_id")
```
Comment (hfcorreia) [inline cmd/dashboard.go]: add to previous statement

File: cmd/root.go [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+	Use:   "logfairy",
+	Short: "logfairy to standardize interaction with graylog",
+	Long:  `logfairy goal is to standardize the way streams, dashboard and widget are created`,
+}
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
```
Comment (hfcorreia) [inline cmd/root.go]: fatal

File: cmd/root.go [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+// initConfig reads in config file and ENV variables if set.
+func initConfig() {
+	viper.SetConfigFile("./graylog.yaml")
+	viper.SetConfigType("yaml")
+
+	viper.AutomaticEnv() // read in environment variables that match
+
+	// If a config file is found, read it in.
+	if err := viper.ReadInConfig(); err != nil {
+		os.Exit(1)
```
Comment (hfcorreia) [inline cmd/root.go]: fatal

File: cmd/root.go [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	service = srv.New(viper.GetString("client.base_url"), viper.GetInt("client.timeout"))
+	graylog = api.New(service, os.Getenv("GRAYLOG_USERNAME"), os.Getenv("GRAYLOG_PASSWORD"))
+}
+
+func prettyPrint(object interface{}) {
+	prettyObject, err := json.MarshalIndent(object, "", "  ")
+	if err != nil {
+		fmt.Println(err)
```
Comment (hfcorreia) [inline cmd/root.go]: missing a return here. you will get 2 prints

File: cmd/stream.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+		prettyPrint(stream)
+	},
+}
+
+// CreateStreamCmd represents the create command
+var CreateStreamCmd = &cobra.Command{
+	Use:   "create",
+	Short: "crate a stream",
+	Long: `stream create will try to create the stream defined in the config file.
```
Comment (hfcorreia) [inline cmd/stream.go]: extract

File: cmd/stream.go [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+		}
+
+		fmt.Printf("stream id: %s", streamID)
+	},
+}
+
+func init() {
+	streamClient = api.NewStreamClient(graylog)
+
+	GetStreamCmd.Flags().StringVarP(&streamID, "stream_id", "s", "", "id of stream to find")
```
Comment (hfcorreia) [inline cmd/stream.go]: format and joint stmt

File: cmd/stream.go [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+	GetStreamCmd.MarkFlagRequired("stream_id")
+
+	CreateStreamCmd.Flags().StringVarP(&definitions, "config", "c", "", "config file containing the stream definition")
```
Comment (hfcorreia) [inline cmd/stream.go]: same as above

File: cmd/widget.go [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+
+		prettyPrint(widget)
+	},
+}
+
+// CreateWidgetCmd represents the create command
+var CreateWidgetCmd = &cobra.Command{
+	Use:   "create",
+	Short: "crate a widget",
+	Long: `widget create will try to create the widget defined in the config file.
```
Comment (hfcorreia) [inline cmd/widget.go]: extract

File: cmd/widget.go [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+	},
+}
+
+func init() {
+	initConfig()
+
+	widgetClient = api.NewWidgetClient(graylog)
+	dashboardClient = api.NewDashboardClient(graylog)
+
+	GetWidgetCmd.Flags().StringVarP(&dashboardWidgetID, "dashboard_id", "d", "", "id of dashboard containing the widget")
```
Comment (hfcorreia) [inline cmd/widget.go]: format and statement

File: infrastructure/api/dashboard.go [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+
+// endpoint constants
+const (
+	dashboards = "/api/dashboards"
+	// /api/dashboards/<deashboard_id>
+	singleDashboard    = "/api/dashboards/%s"
+	dashboardTokenName = "dashboards"
+)
+
+type DashboardClient struct {
```
Comment (hfcorreia) [inline infrastructure/api/dashboard.go]: missing a comment. do we have linting for this project?

File: infrastructure/api/dashboard.go [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+	Graylog
+}
+
+// New create an instance of Graylog deashboard api client
+func NewDashboardClient(graylog Graylog) DashboardClient {
+	return DashboardClient{Graylog: graylog}
+}
+
+func (dashboardClient *DashboardClient) List() (dashboard.Dashboards, error) {
```
Comment (hfcorreia) [inline infrastructure/api/dashboard.go]: comment

File: infrastructure/api/dashboard.go [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+		return dashboard.Dashboard{}, err
+	}
+
+	endpoint := fmt.Sprintf(singleDashboard, dashboardID)
+	response, status, err := dashboardClient.client.Get(endpoint, headers, auth, nil)
+	if err != nil {
+		return dashboard.Dashboard{}, err
+	}
+
+	failure, err := dashboardClient.handleFailure(response, status)
```
Comment (hfcorreia) [inline infrastructure/api/dashboard.go]: why does this return the failure? why not wrap it in the err?

File: infrastructure/api/dto/token.go [hunk 28]
```diff
@@ -0,0 +1,7 @@
+package dto
+
+type TokenResponse struct {
```
Comment (hfcorreia) [inline infrastructure/api/dto/token.go]: comment

File: infrastructure/api/graylog.go [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/uniplaces/logfairy/infrastructure/api/dto"
+	"github.com/uniplaces/logfairy/infrastructure/api/service"
+)
+
+// endpoint constants
+const (
+	// users/<username>/tokens/<token_name>
```
Comment (hfcorreia) [inline infrastructure/api/graylog.go]: improve this comment.

File: infrastructure/api/graylog.go [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+	tokensStream = "/api/users/%s/tokens/%s"
+)
+
+var headers = map[string]string{
+	"Content-Type": "application/json",
+}
+
+type Graylog struct {
```
Comment (hfcorreia) [inline infrastructure/api/graylog.go]: comment

File: infrastructure/api/graylog.go [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+	}
+}
+
+func (graylog *Graylog) getAuth(tokenName string) ([]string, error) {
+	token, err := graylog.getToken(tokenName)
+	if err != nil {
+		return []string{}, err
+	}
+
+	return []string{token, "token"}, nil
```
Comment (hfcorreia) [inline infrastructure/api/graylog.go]: this seems weird, why do you need the "token"?

File: infrastructure/api/graylog.go [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+		return err
+	}
+
+	graylog.token = tokenResponse.Token
+
+	return nil
+}
+
+func (graylog *Graylog) handleFailure(response []byte, status int) (*dto.ErrorResponse, error) {
+	if status >= 200 && status < 299 {
```
Comment (hfcorreia) [inline infrastructure/api/graylog.go]: use the http.StatusOK?

File: infrastructure/api/service/service.go [hunk 33]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	return body, resp.StatusCode, nil
+}
+
+func (service Service) buildURL(endpoint string) string {
+	return service.BaseURL + endpoint
+}
+
+func TimeoutDialer(cTimeout time.Duration) func(net, addr string) (c net.Conn, err error) {
```
Comment (hfcorreia) [inline infrastructure/api/service/service.go]: comment

File: infrastructure/api/stream.go [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+	"fmt"
+
+	"github.com/uniplaces/logfairy/dto/stream"
+	"github.com/uniplaces/logfairy/infrastructure/api/dto"
+)
+
+// endpoint constants
+const (
+	streams = "/api/streams"
+	// /api/streams/<stream_id>
```
Comment (hfcorreia) [inline infrastructure/api/stream.go]: improve comment 💃

File: infrastructure/api/stream.go [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+	singleStream    = "/api/streams/%s"
+	streamTokenName = "streams"
+)
+
+type StreamClient struct {
```
Comment (hfcorreia) [inline infrastructure/api/stream.go]: comment

File: infrastructure/api/stream.go [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+	Graylog
+}
+
+// New create an instance of Graylog stream api client
+func NewStreamClient(graylog Graylog) StreamClient {
+	return StreamClient{Graylog: graylog}
+}
+
+func (streamClient *StreamClient) List() (stream.Streams, error) {
```
Comment (hfcorreia) [inline infrastructure/api/stream.go]: comment

File: infrastructure/api/stream.go [hunk 37]
```diff
@@ -0,0 +1,115 @@
...
+
+	list := stream.Streams{}
+	if err := json.Unmarshal(response, &list); err != nil {
+		return stream.Streams{}, err
+	}
+
+	return list, nil
+}
+
+func (streamClient *StreamClient) Get(streamID string) (stream.Stream, error) {
```
Comment (hfcorreia) [inline infrastructure/api/stream.go]: comment

File: infrastructure/api/stream.go [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+
+	streamFound := stream.Stream{}
+	if err := json.Unmarshal(response, &streamFound); err != nil {
+		return stream.Stream{}, err
+	}
+
+	return streamFound, nil
+}
+
+func (streamClient *StreamClient) Create(streamToCreate stream.Stream) (string, error) {
```
Comment (hfcorreia) [inline infrastructure/api/stream.go]: comment

File: infrastructure/api/widget.go [hunk 39]
```diff
@@ -0,0 +1,90 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/uniplaces/logfairy/dto/dashboard"
+	"github.com/uniplaces/logfairy/infrastructure/api/dto"
+)
+
+// endpoint constants
+const (
```
Comment (hfcorreia) [inline infrastructure/api/widget.go]: improve comment

File: infrastructure/api/widget.go [hunk 40]
```diff
@@ -0,0 +1,90 @@
...
+	// /api/dashboards/<dashboard_id>/widgets/<widget_id>
+	singleWidget = "/api/dashboards/%s/widgets/%s"
+	// /api/dashboards/<dashboard_id>/widgets
+	createWidget    = "/api/dashboards/%s/widgets"
+	widgetTokenName = "widgets"
+)
+
+type WidgetClient struct {
```
Comment (hfcorreia) [inline infrastructure/api/widget.go]: comment

File: infrastructure/api/widget.go [hunk 41]
```diff
@@ -0,0 +1,90 @@
...
+	Graylog
+}
+
+// New create an instance of Graylog deashboard api client
+func NewWidgetClient(graylog Graylog) WidgetClient {
+	return WidgetClient{Graylog: graylog}
+}
+
+func (widgetClient *WidgetClient) Get(widgetID string, dashnoardID string) (dashboard.Widget, error) {
```
Comment (hfcorreia) [inline infrastructure/api/widget.go]: comment

File: infrastructure/api/widget.go [hunk 42]
```diff
@@ -0,0 +1,90 @@
...
+
+	dashboardFound := dashboard.Widget{}
+	if err := json.Unmarshal(response, &dashboardFound); err != nil {
+		return dashboard.Widget{}, err
+	}
+
+	return dashboardFound, nil
+}
+
+func (widgetClient *WidgetClient) Create(widgetToCreate dashboard.Widget, dashnoardID string) (string, error) {
```
Comment (hfcorreia) [inline infrastructure/api/widget.go]: comment

File: infrastructure/api/graylog.go [hunk 43]
```diff
@@ -0,0 +1,99 @@
...
+	if err := json.Unmarshal(response, tokenResponse); err != nil {
+		return err
+	}
+
+	graylog.token = tokenResponse.Token
+
+	return nil
+}
+
+func (graylog *Graylog) handleFailure(response []byte, status int) (*dto.ErrorResponse, error) {
```
Comment (hfcorreia) [inline infrastructure/api/graylog.go]: don't return *dto.ErrorResponse, just wrap it in an error

File: cmd/bulk.go (hunk 9 above)
Comment (raffo0707) [reply cmd/bulk.go]: yes, graylog api doesn't accept the widget, I read the widgets from the dashboard to be easier to define. But then I create one by one

File: cmd/stream.go (hunk 21 above)
Comment (raffo0707) [reply cmd/stream.go]: 🇯🇲

File: infrastructure/api/graylog.go (hunk 31 above)
Comment (raffo0707) [reply infrastructure/api/graylog.go]: is to specify that the basic auth is using a token, is in the doc

File: infrastructure/api/service/service.go (hunk 33 above)
Comment (raffo0707) [reply infrastructure/api/service/service.go]: this should be private

**Pull request merged**